src/gwt/src/org/rstudio/studio/client/common/shell/ShellWidget.java

```diff
@@ -457,24 +454,60 @@ private boolean trimExcess()
 
    public void playbackActions(final RpcObjectList<ConsoleAction> actions)
    {
+      // Server persists 1000 most recent ConsoleActions in a circular buffer.
+      //
+      // One ConsoleAction can generate multiple lines of output, and we want 
+      // to limit number of lines added to the console's DOM; see trimExcess().
+      //
+      // First walk through the actions in reverse, and determine how many 
+      // lines they will generate (without actually writing anything),
+      // then play-back in normal order. Finally, trim to the max-lines we support
+      // to catch any rounding from final chunk.
+      int lines = 0;
+      int excessLines = 0;
+      int revIndex = actions.length() - 1;
+      for (; revIndex >= 0; revIndex--)
+      {
+         ConsoleAction action = actions.get(revIndex);
+         
+         if (action.getType() == ConsoleAction.INPUT)
+            lines++;
+         
+         lines = lines + StringUtil.newlineCount(action.getData());
+         
+         excessLines = excessLines(lines);
```

Comment:
nit: could we just omit the excessLines variable and function altogether and just say `if (lines > maxLines_)` for simplicity?